File: bench/stress/s7_align.go
// S7 — local pairwise alignment, Go (std only). Score-only Gotoh affine DP,
// replicating src/align.rs local mode. go run s7_align.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const (
	NEG   = -(1 << 30)
	MATCH = 1
	MIS   = -1
	GEXT  = -1
	FG    = -6 // gap_open + gap_extend
)

func max2(a, b int32) int32 {
	if a > b {
		return a
	}
	return b
}

func localScore(x, y []byte) int32 {
	m := len(y)
	prevMM := make([]int32, m+1)
	prevXX := make([]int32, m+1)
	prevYY := make([]int32, m+1)
	for j := range prevMM {
		prevMM[j], prevXX[j], prevYY[j] = NEG, NEG, NEG
	}
	prevMM[0] = 0
	for j := 1; j <= m; j++ {
		prevYY[j] = max2(prevMM[j-1]+FG, prevYY[j-1]+GEXT)
	}
	var best int32 = 0
	curMM := make([]int32, m+1)
	curXX := make([]int32, m+1)
	curYY := make([]int32, m+1)
	for _, xi := range x {
		curMM[0], curXX[0], curYY[0] = NEG, max2(prevMM[0]+FG, prevXX[0]+GEXT), NEG
		for j := 1; j <= m; j++ {
			var s int32 = MIS
			if xi == y[j-1] {
				s = MATCH
			}
			d := max2(max2(prevMM[j-1], prevXX[j-1]), prevYY[j-1])
			mval := d + s
			if mval < 0 {
				mval = 0
			}
			curMM[j] = mval
			if mval > best {
				best = mval
			}
			curXX[j] = max2(prevMM[j]+FG, prevXX[j]+GEXT)
			curYY[j] = max2(curMM[j-1]+FG, curYY[j-1]+GEXT)
		}
		prevMM, curMM = curMM, prevMM
		prevXX, curXX = curXX, prevXX
		prevYY, curYY = curYY, prevYY
	}
	return best
}

func readFasta(path string) [][]byte {
	f, _ := os.Open(path)
	defer f.Close()
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 1<<20), 1<<20)
	var seqs [][]byte
	var cur []byte
	for sc.Scan() {
		line := sc.Bytes()
		if len(line) > 0 && line[0] == '>' {
			if len(cur) > 0 {
				seqs = append(seqs, cur)
				cur = nil
			}
		} else {
			cur = append(cur, line...)
		}
	}
	if len(cur) > 0 {
		seqs = append(seqs, cur)
	}
	return seqs
}

func main() {
	reference := readFasta("data/ref.fa")[0]
	var total, hits int64
	var max int32
	for _, q := range readFasta("data/queries.fa") {
		s := localScore(q, reference)
		total += int64(s)
		if s >= 40 {
			hits++
		}
		if s > max {
			max = s
		}
	}
	fmt.Printf("total=%d\n", total)
	fmt.Printf("hits=%d\n", hits)
	fmt.Printf("max=%d\n", max)
}

File: bench/stress/s7_align.rs
// S7 — local pairwise alignment, Rust (std only). Score-only Gotoh affine DP,
// replicating src/align.rs local mode. rustc -O s7_align.rs -o s7_align
use std::fs;

const NEG: i32 = i32::MIN / 4;
const MATCH: i32 = 1;
const MIS: i32 = -1;
const GEXT: i32 = -1;
const FG: i32 = -6; // gap_open + gap_extend

fn local_score(x: &[u8], y: &[u8]) -> i32 {
    let m = y.len();
    let mut prev_mm = vec![NEG; m + 1];
    let mut prev_xx = vec![NEG; m + 1];
    let mut prev_yy = vec![NEG; m + 1];
    prev_mm[0] = 0;
    for j in 1..=m {
        prev_yy[j] = (prev_mm[j - 1] + FG).max(prev_yy[j - 1] + GEXT);
    }
    let mut best = 0;
    let mut cur_mm = vec![NEG; m + 1];
    let mut cur_xx = vec![NEG; m + 1];
    let mut cur_yy = vec![NEG; m + 1];
    for &xi in x {
        cur_mm[0] = NEG;
        cur_xx[0] = (prev_mm[0] + FG).max(prev_xx[0] + GEXT);
        cur_yy[0] = NEG;
        for j in 1..=m {
            let s = if xi == y[j - 1] { MATCH } else { MIS };
            let d = prev_mm[j - 1].max(prev_xx[j - 1]).max(prev_yy[j - 1]);
            let mut mval = d + s;
            if mval < 0 {
                mval = 0;
            }
            cur_mm[j] = mval;
            if mval > best {
                best = mval;
            }
            cur_xx[j] = (prev_mm[j] + FG).max(prev_xx[j] + GEXT);
            cur_yy[j] = (cur_mm[j - 1] + FG).max(cur_yy[j - 1] + GEXT);
        }
        std::mem::swap(&mut prev_mm, &mut cur_mm);
        std::mem::swap(&mut prev_xx, &mut cur_xx);
        std::mem::swap(&mut prev_yy, &mut cur_yy);
    }
    best
}

fn read_fasta(path: &str) -> Vec<Vec<u8>> {
    let raw = fs::read_to_string(path).unwrap();
    let mut seqs = Vec::new();
    let mut cur: Vec<u8> = Vec::new();
    for line in raw.lines() {
        if line.starts_with('>') {
            if !cur.is_empty() {
                seqs.push(std::mem::take(&mut cur));
            }
        } else {
            cur.extend(line.bytes());
        }
    }
    if !cur.is_empty() {
        seqs.push(cur);
    }
    seqs
}

fn main() {
    let ref_seqs = read_fasta("data/ref.fa");
    let reference = &ref_seqs[0];
    let mut total: i64 = 0;
    let mut hits: i64 = 0;
    let mut max = 0;
    for q in read_fasta("data/queries.fa") {
        let s = local_score(&q, reference);
        total += s as i64;
        if s >= 40 {
            hits += 1;
        }
        if s > max {
            max = s;
        }
    }
    println!("total={}", total);
    println!("hits={}", hits);
    println!("max={}", max);
}
